### PR TITLE
Category expansion breaks Beautifier

### PR DESCRIPTION
**Pull request by groupby-r2d2**
Created by: ferronrsmith 
Parent groupby/issues#920

If an expanded category is used as a Navigation, the refinement breaks the beautifier.

```
category3_expanded: Category Root~Athletics~Boys'
```

The beautifier breaks because the category contains tilde (~), which is used as a delimiter.
- [ ] groupby/bindle#1533

### CODE REVIEW

**Comment by ferronrsmith** [comment]
@osman can you review this, thanks.

**Comment by osman** [comment]
@ferronrsmith looks good to me, we should do a demo tomorrow and get this merged.

File: common/src/main/java/com/groupbyinc/api/AbstractQuery.java [hunk 1]
```diff
@@ -353,7 +353,7 @@ public Q addRefinementsByString(String refinementString) {
         if (refinementString == null) {
             return (Q) this;
         }
-        String[] filterStrings = refinementString.split("~");
+        String[] filterStrings = refinementString.split("~((?=[\\w]*[=:]))");
```
Comment (codeallthethingz) [inline common/src/main/java/com/groupbyinc/api/AbstractQuery.java]: Can you add a java comment here that describes the behaviour of the regex so that those who come later might understand?

File: flux/src/test/java/com/groupbyinc/util/UrlFunctionsBugTest.java [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+import static org.junit.Assert.assertArrayEquals;
+import static org.junit.Assert.assertEquals;
+
+/**
+ * Created by ferron on 4/22/15.
+ */
+public class UrlFunctionsBugTest {
+
+    private final String DEFAULT_BEAUTIFIER = "default";
+    private final String REFINEMENT_REGEX = "~((?=[\\w]*[=:]))";
```
Comment (codeallthethingz) [inline flux/src/test/java/com/groupbyinc/util/UrlFunctionsBugTest.java]: This should be defined in the AbstractQuery class and referenced here.  Keep it DRY.

File: flux/src/test/java/com/groupbyinc/util/UrlFunctionsBugTest.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+        String url = UrlFunctions.toUrlAdd(
+                DEFAULT_BEAUTIFIER, "", navigations, "category_leaf_expanded", new RefinementValue().setValue(
+                        refinementString).setCount(5483));
+        assertEquals(
+                "/index.html?refinements=%7Ecategory_leaf_expanded%3DCategory+Root%7EAthletics%7EMen%27s%7ESneakers",
+                url);
+    }
+
+    @Test
```
Comment (codeallthethingz) [inline flux/src/test/java/com/groupbyinc/util/UrlFunctionsBugTest.java]: This test actually tests REGEX, not the class UrlBeautifier.  If someone changes the class UrlBeautifier this test will never break and is useless from protecting us against other developers.

I think you should add a method to the UrlBeautifier that does the split logic, make it package private and have this test call that method.

File: flux/src/test/java/com/groupbyinc/util/UrlFunctionsBugTest.java [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+        String[] split = refinementGood.split(REFINEMENT_REGEX);
+
+        assertArrayEquals(new String[]{"", "gender=Women", "simpleColorDesc=Pink", "product=Clothing"}, split);
+
+        String refinementBad = "~category_leaf_expanded=Category Root~Athletics~Men's~Sneakers";
+
+        split = refinementBad.split(REFINEMENT_REGEX);
+
+        assertArrayEquals(
```
Comment (codeallthethingz) [inline flux/src/test/java/com/groupbyinc/util/UrlFunctionsBugTest.java]: Test methods should test one thing and have a descriptive name.  You're testing lots and lots of things here.  

I would say you need methods called

``` java
splitTestExpected()
splitTestCategory()
splitTestCategoryLong()
etc...
```

File: flux/src/test/java/com/groupbyinc/util/UrlFunctionsBugTest.java [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+                new String[]{
+                        "", "category_leaf_expanded=Category Root~Athletics~Men's~Sneakers", "category_leaf_id=580003",
+                        "color=BLUE", "color=YELLOW", "color=GREY", "feature=Lace Up", "feature=Light Weight",
+                        "brand=Nike"
+                },
+                split);
+    }
+}
```
Comment (codeallthethingz) [inline flux/src/test/java/com/groupbyinc/util/UrlFunctionsBugTest.java]: You need tests that test the edge conditions for the method call that you're now going to create.

``` java
testNull()
testEmpty()
testUtf8()  // Remember that *any* test that tests a string must have an umlaut in at least one of the tests.
```

**Comment by codeallthethingz** [comment]
@ferronrsmith lots of changes.

@osman Code review is a great time to make your developers better and stronger and faster.  If you feel you don't have time to do a code review properly come talk to me and we will try and fix the underlying issues.

**Comment by codeallthethingz** [comment]
@ferronrsmith does this split change neei to be in the PHP api too?

**Comment by ferronrsmith** [comment]
@willwarren yes the php api needs to be updated.
